### PR TITLE
Closes #656 Add option to use image aspect ratio for background aspect ratio.

### LINKED ISSUE #656: Text on Media paragraph type updates
**Requirements:**
- [x] #893
- [x]  #1024  
- [x]  #995 Allow bottom style media to be cropped/adjusted 
- [ ] ~#1039~
- [x] #1041
- [x] #1044
- [x] #1028  
**Additional context:**
Will need to ensure option updates are applied to both image and video backgrounds

### PR DESCRIPTION
**Pull request by trackleft**
## How Has This Been Tested?
Go to /node/14/edit
Change text on media paragraph space around content to 'Background Aspect Ratio'
Change crops for each responsive background style
Save
See override of existing background size options.

**Note: The 'Background Aspect Ratio' option will be available even if the full width option is not checked, but shouldn't do anything.**

Go to /node/16/edit and repeat for all text on media paragraphs.

Change any full width paragraph
## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [x] I've added this Pull Request to the AZ Quickstart Github project and set it to "Review in progress".
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: modules/custom/az_paragraphs/css/az_paragraphs_az_text_media.css [hunk 1]
```diff
@@ -26,26 +26,19 @@
...
-}
-.media--type--az-remote-video.box {
-  padding: 0;
-}
 .az-video-playing,
 .az-video-playing .text-on-video {
   background: none !important;
 }
-.column {
+.paragraph--type--az-text-media .box {
```
Comment (trackleft) [inline modules/custom/az_paragraphs/css/az_paragraphs_az_text_media.css:33]: this might not be what we want.

**Comment by trackleft** [comment]
Looks like column view doesn't act like a column

**Comment by trackleft** [comment]
Sidebars doesn't work that well either.

**Comment by smayden** [comment]
This is a great start! Initial testing of the cropping functionality/display of associated crops seems to be working well. Noticing problems with the placement for both column and box (e.g. offset to the left)

**Comment by camikazegreen** [comment]
- [ ] Box and Column style position **left spacing** goes all the way to the edge
- [ ] Box and Column style position **none spacing** goes all the way to the edge
- [ ] Column style is not a column
- [ ] Bottom style is not on the bottom

Confirmed that all other iterations work as expected, and that the existing "Padding around Content" settings do not have the listed display issues. (Only tested on desktop)

**Comment by trackleft** [comment]
During the process of creating this PR, I did notice some issues with the width of the column or box when there is a sidebar.

File: modules/custom/az_paragraphs/az_paragraphs_text_media/az_paragraphs_text_media.module [hunk 2]
```diff
@@ -8,7 +8,7 @@
 /**
  * Implements hook_theme().
  */
-function az_paragraphs_text_media_theme($existing, $type, $theme, $path) {
+function az_paragraphs_text_media_theme($existing, $type, $theme, $path): array {
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/az_paragraphs_text_media/az_paragraphs_text_media.module:11]: FWIW, I think it'd probably be better to add return type hints to all functions/methods in a separate issue/PR, maybe as part of preparing for Drupal 9.3.x/10.0.x or PHP 8 but maybe that's just me.

Comment (trackleft) [reply modules/custom/az_paragraphs/az_paragraphs_text_media/az_paragraphs_text_media.module:11]:
```suggestion
function az_paragraphs_text_media_theme($existing, $type, $theme, $path) {
```

File: modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZTextWithMediaParagraphBehavior.php [hunk 3]
```diff
@@ -139,6 +139,7 @@ public function buildBehaviorForm(ParagraphInterface $paragraph, array &$form, F
         'y-10' => $this->t('10 (8.0rem | ~128px)'),
         'y-20' => $this->t('20 (16.0rem | ~256px)'),
         'y-30' => $this->t('30 (24.0rem | ~384px)'),
+        'az-aspect-ratio' => $this->t('Background Aspect Ratio'),
```
Comment (danahertzberg) [inline modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZTextWithMediaParagraphBehavior.php:142]:
```suggestion
        'az-aspect-ratio' => $this->t('Media Aspect Ratio'),
```

File: modules/custom/az_paragraphs/src/Plugin/Field/FieldFormatter/AZBackgroundMediaFormatter.php [hunk 4]
```diff
@@ -512,7 +512,7 @@ protected function remoteVideo(array $settings, MediaInterface $media): array {
         '#attributes' => [
           'id' => [$video_oembed_id . '-bg-video-container'],
           'class' => [
-            'az-video-loading',
+            // 'az-video-loading',
```
Comment (trackleft) [inline modules/custom/az_paragraphs/src/Plugin/Field/FieldFormatter/AZBackgroundMediaFormatter.php:515]:
```suggestion
```

File: modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphsBehavior.php [hunk 5]
```diff
@@ -166,7 +166,7 @@ public function view(array &$build, Paragraph $paragraph, EntityViewDisplayInter
 
     // Apply bottom spacing if set.
     if (!empty($config['az_display_settings']['bottom_spacing'])) {
-      $build['#attributes']['class'] = $config['az_display_settings']['bottom_spacing'];
+      $build['#attributes']['class'] = [$config['az_display_settings']['bottom_spacing']];
```
Comment (trackleft) [inline modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphsBehavior.php:169]:
```suggestion
      $build['#attributes']['class'] = $config['az_display_settings']['bottom_spacing'];
```

**Review by smayden — APPROVED** [review]
Tested front end comparing text on media with different style configurations with and without the media aspect ratio. Cropping is working well and the display using media aspect ratio is all looking good now. Impressive!

The only problem I had was videos appearing much taller than expected when no thumbnail is added (worked as expected with thumbnail). I think this could be addressed as a follow up.

**Review by joeparsons — APPROVED** [review]
This looks good to me code-wise.

**Review by tadean — APPROVED** [review]

**Pull request merged**